add fetch billing products from tables instead of env variables

anamarn opened the pull request:
Solves https://github.com/twentyhq/private-issues/issues/237

**TLDR:**

- Fetches billing products and prices from the tables BilllingProducts and BillingPrices instead of fetching the product from the environment variables and the prices from the stripe API.
- Adds new feature flag for this feature
- Fixes calls used to fetch stripe products and prices for the command Billing Sync Plans Data. 


**In order to test:**

1. Have the environment variable IS_BILLING_ENABLED set to true and add the other required environment variables for Billing to work
2. Do a database reset (to ensure that the new feature flag is properly added and that the billing tables are created)
3.  Run the command: `npx nx run twenty-server:command billing:sync-plans-data` (if you don't do that the products and prices will not be present in the database)
4. Run the server , the frontend, the worker, and the stripe listen command (`stripe listen --forward-to http://localhost:3000/billing/webhooks`)
5. Buy a subscription for the Acme workspace and play with the project

**Doing**

I think there is some room of progress for the function formatProductPrices, I used a similar version that was done before, I'll look into that.

Weiko commented on packages/twenty-server/src/engine/core-modules/billing/billing.resolver.ts:
```diff
@@ -24,16 +29,39 @@ export class BillingResolver {
+    private readonly featureFlagService: FeatureFlagService,
   ) {}
 
+  async getBillingPllansEnabledFeatureFlag(workspaceId: string) {
```
typo

Weiko commented on packages/twenty-server/src/engine/core-modules/billing/billing.resolver.ts:
```diff
@@ -24,16 +29,39 @@ export class BillingResolver {
   ) {}
 
+  async getBillingPllansEnabledFeatureFlag(workspaceId: string) {
+    return await this.featureFlagService.isFeatureEnabled(
```
this method returns a bool, I think you can rename yours `isBillingPlanEnabled`  to make it clear, this is because you are not returning the flag but its value and usually we prefer to name methods with "is" once returning bools.
Although I don't think this method is bringing much value, up to you!

Weiko commented on packages/twenty-server/src/engine/core-modules/billing/constants/available-product-to-plans-map.constant.ts:
```diff
@@ -0,0 +1,10 @@
+import { BillingPlanKey } from 'src/engine/core-modules/billing/enums/billing-plan-key.enum';
+import { BillingUsageType } from 'src/engine/core-modules/billing/enums/billing-usage-type.enum';
+
+export const AVAILABLE_PRODUCT_TO_PLANS_MAP = {
```
👍

charlesBochet replied:
remove

Weiko commented on packages/twenty-server/src/engine/core-modules/billing/services/billing-plan.service.ts:
```diff
@@ -0,0 +1,92 @@
+    planKey: BillingPlanKey,
+    priceUsageBased: BillingUsageType,
+  ): Promise<string> {
+    const product = await this.billingProductRepository.findOneOrFail({
```
I think it's better to throw a custom Billing exception if no product is found to follow our patterns (you can also try catch here and rethrow, up to you)

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/billing.resolver.ts:
```diff
@@ -24,16 +29,39 @@ export class BillingResolver {
-    const productPrices =
-      await this.stripePriceService.getStripePrices(product);
+  @UseGuards(WorkspaceAuthGuard)
+  async getProductPrices(
```
introduce new endpoint getProduct that will support front usecase

charlesBochet replied:
getProductsByPlan that returns

```
{ 
 proPlan: {
  key,
  baseProduct: { 
    stripeId,
    prices,
    ...
  }
  products: {
  
  }
}
```

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/billing.resolver.ts:
```diff
@@ -64,11 +92,21 @@ export class BillingResolver {
     );
 
+    const productPrice = isBillingPlansEnabled
+      ? await this.billingPlanService.getProductPriceByPlanAndInterval(
```
getBaseProductPriceForPlanAndInterval()

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/services/billing-plan.service.ts:
```diff
@@ -0,0 +1,92 @@
+    private readonly billingProductRepository: Repository<BillingProduct>,
+  ) {}
+
+  async getProductByPlan(
```
getBaseProductFStripeIdorPlan ? is is still useful
or getProductStripeIdsForPlan  ? is is still useful

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/services/billing-subscription.service.ts:
```diff
@@ -10,28 +10,41 @@ import { BillingEntitlement } from 'src/engine/core-modules/billing/entities/bil
     private readonly billingSubscriptionRepository: Repository<BillingSubscription>,
   ) {}
 
+  async getBillingPllansEnabledFeatureFlag(workspaceId: string) {
```
not useful

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/services/billing-subscription.service.ts:
```diff
@@ -56,14 +69,23 @@ export class BillingSubscriptionService {
+        )
+      : stripeProductId;
 
     const billingSubscriptionItem =
```
is this check useful?
I would remove it

anamarn commented:
- Add a new endpoint in graphql  named `getAllPlansInformation()` who let us check and get all the products and prices available for each plan. If you want to test it out I recommend using yoga graphQL.
- Renamed some entities that should be considered dtos.
- Added new methods for the billing plans services and created more generalized methods
-  Introduced the concept of baseProduct to the BillingPlans

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/billing.resolver.ts:
```diff
@@ -87,11 +117,17 @@ export class BillingResolver {
   }
+
+  @Query(() => PlansInformationDTO)
+  // @UseGuards(WorkspaceAuthGuard) I don't need the workspace for this query should i add it anyways?
```
leave guard activated

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/billing.resolver.ts:
```diff
@@ -24,20 +28,33 @@ export class BillingResolver {
+    @AuthWorkspace() workspace: Workspace,
+    @Args() { product }: ProductInput,
+  ) {
+    const isBillingPlansEnabled =
```
revert change as it was before

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/dto/plan-information.dto.ts:
```diff
@@ -0,0 +1,19 @@
+import { BillingPlanKey } from 'src/engine/core-modules/billing/enums/billing-plan-key.enum';
+
+@ObjectType()
+export class PlanInformationDTO {
```
BillingPlanDTO
Add billing every where

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/dto/update-billing.dto.ts:
```diff
@@ -1,7 +1,7 @@
 
 @ObjectType()
-export class UpdateBillingEntity {
+export class UpdateBillingDTO {
```
/dtos/outpus/update-subscription.output.ts (UpdateSubscriptionOutput)

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/dto/billing-price-tier.dto.ts:
```diff
@@ -0,0 +1,13 @@
+import { Field, ObjectType } from '@nestjs/graphql';
```
BillingPriceTierDTO

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/billing.resolver.ts:
```diff
@@ -87,11 +117,17 @@ export class BillingResolver {
+
+  @Query(() => PlansInformationDTO)
+  // @UseGuards(WorkspaceAuthGuard) I don't need the workspace for this query should i add it anyways?
+  async getAllPlansInformation() {
```
plans

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/dto/plans-information.dto.ts:
```diff
@@ -0,0 +1,9 @@
+import { Field, ObjectType } from '@nestjs/graphql';
```
in outputs
plans.output.ts

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/dto/price-licensed.dto.ts:
```diff
@@ -0,0 +1,15 @@
+import { SubscriptionInterval } from 'src/engine/core-modules/billing/enums/billing-subscription-interval.enum';
+
+@ObjectType()
+export class PriceLicensedDTO {
```
BillingPrice

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/services/billing-plan.service.ts:
```diff
@@ -0,0 +1,268 @@
+  ) {}
+
+  async getProductsByProductMetadata(
+    planKey: BillingPlanKey,
```
1. use object as soon as you have more than one parameter
getProductsByProductMetadata({ planKey, priceUsageBased, isBaseProduct}: {{ planKey: BillingPlanKey, priceUsageBased: BillingUsageType, isBaseProduct: 'true' | 'false })

2. BillingPriceUsageType : priceUsageType
3. isBaseProduct: boolean

anamarn replied:
I will do the points 2 and 3 on another PR , creating the corresponding issue right now

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/services/billing-plan.service.ts:
```diff
@@ -0,0 +1,268 @@
+    return products;
+  }
+
+  async getAllPlansInformation(): Promise<PlansInformationDTO> {
```
getAllPlans()

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/services/billing-plan.service.ts:
```diff
@@ -0,0 +1,268 @@
+
+    const plans = await Promise.all(
+      planKeys.map(async (planKey) => {
+        return await this.getPlanInformationByKey(planKey);
```
getPlanByKey

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/services/billing-subscription.service.ts:
```diff
@@ -56,19 +62,35 @@ export class BillingSubscriptionService {
     if (!billingSubscriptionItem) {
       throw new Error(
-        `Cannot find billingSubscriptionItem for product ${stripeProductId} for workspace ${workspaceId}`,
+        `Cannot find billingSubscriptionItem for product ${getStripeProductId} for workspace ${workspaceId}`,
```
remove this logic

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/services/billing-subscription.service.ts:
```diff
@@ -136,10 +162,17 @@ export class BillingSubscriptionService {
-    );
+    const productPrice = isBillingPlansEnabled
+      ? (
+          await this.billingPlanService.getProductPrices(
```
keep

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/services/billing-plan.service.ts:
```diff
@@ -0,0 +1,268 @@
+    );
+  }
+
+  async getProductPrices(
```
remove

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/services/billing-plan.service.ts:
```diff
@@ -0,0 +1,268 @@
+    };
+  }
+
+  formatBillingPriceDataToProductPriceDTO(
```
move this to the resolver

charlesBochet replied:
or to utils

charlesBochet commented on packages/twenty-server/src/database/typeorm/core/migrations/billing/1737127856478-addNonNullableProductDescription.ts:
```diff
@@ -0,0 +1,25 @@
+  name = 'AddNonNullableProductDescription1737127856478';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
```
agree with Coco, or we should do both in same instruction

charlesBochet replied:
checking on our production database if we have billingProduct with description NULL: we don't so it will work

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/billing.resolver.ts:
```diff
@@ -87,11 +121,43 @@ export class BillingResolver {
+      return {
+        planKey: plan.planKey,
+        baseProduct: {
+          ...plan.baseProduct,
```
create an util:
- formatDatabaseProductToGraphqlDTO

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/dtos/billing-product.dto.ts:
```diff
@@ -0,0 +1,24 @@
+  @Field(() => String)
+  description: string;
+
+  @Field(() => [String], { nullable: true })
```
{ nullable: false } default: ''

charlesBochet commented on packages/twenty-server/src/engine/core-modules/billing/utils/format-billing-price-data-to-licensed-price-dto.util copy.ts:
```diff
@@ -0,0 +1,13 @@
+import { BillingPrice } from 'src/engine/core-modules/billing/entities/billing-price.entity';
+import { SubscriptionInterval } from 'src/engine/core-modules/billing/enums/billing-subscription-interval.enum';
+
+export const formatBillingPriceDataToLicensedPriceDTO = (
```
filename

charlesBochet reviewed (approved):
Bravo!
